### ISSUE TITLE
How to use DetMetrics() and DetectionValidator() for offline AP calculation

### ISSUE THREAD

**Issue by cavalier501** [ultralytics/ultralytics#22121]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I have saved a set of model detection predictions and ground‑truth labels and would like to compute their AP offline using the Ultralytics code. I saw that you mentioned an offline AP calculation method in [issue](https://github.com/ultralytics/ultralytics/issues/19057#issuecomment-2636314477), but due to a recent version update, running that demo directly now raises an error (specifically: `_process_batch` receives arguments in the wrong format; in the current version the argument should be a dictionary containing predictions and labels).

My Ultralytics version is **8.3.174**. Could you please provide a demo that works with the newer version?

Also, I noticed that the demo calls `metrics.process()` after each image is processed. Since AP needs to be computed from the aggregated predictions over the entire dataset (rather than simply summing APs from individual images), I would like to confirm: does the demo compute AP only after all images have been processed (i.e., after `metrics.results_dict` is available)?

Thank you for your help much.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @cavalier501, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get support faster. An Ultralytics engineer will review your issue and assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. For your case, please include:
-1–3 minimal sample predictions and labels showing your current input structure (saved files or in-memory dicts)  
-The exact code snippet calling `DetMetrics()` / `DetectionValidator()` and the point where `_process_batch` is invoked  
-Full error traceback  
-Your environment details: OS, CUDA/CUDNN, Python, PyTorch, and `ultralytics` version  
-The output of `python -c "import ultralytics, torch; print(ultralytics.__version__, torch.__version__)"`  
This will help us verify any API changes and reproduce the format mismatch you’re seeing.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

Tip: After upgrading, please re-run your offline metrics script to confirm whether the `_process_batch` signature issue persists, and share any updated logs 📋.

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for the detailed report and version info (8.3.174) 🙏. With the MRE and logs, we can quickly confirm the expected dictionary format and aggregation behavior for AP in the current release and guide you accordingly.

**Comment by Y-T-G** [comment]
I updated it:

https://github.com/ultralytics/ultralytics/issues/19057#issuecomment-2636314477

**Comment by cavalier501** [comment]
> I updated it:
> 
> [#19057 (comment)](https://github.com/ultralytics/ultralytics/issues/19057#issuecomment-2636314477)

Thanks for your reply much.

**Comment by glenn-jocher** [comment]
Glad it helped—confirming: AP is computed once at the end when you call `DetMetrics.process()`, which aggregates all per-image stats; you can then read `metrics.results_dict` or `metrics.mean_results()` for mp, mr, mAP50, mAP50-95. See the method details in the [DetMetrics reference](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.DetMetrics.process) and how stats are populated in [DetectionValidator.update_metrics](https://docs.ultralytics.com/reference/models/yolo/detect/val/#ultralytics.models.yolo.detect.val.DetectionValidator.update_metrics).

Here’s a minimal offline example for 8.3.174 using your saved preds/labels; ensure GT and prediction boxes are in the same coordinate frame (XYXY), and classes are zero-indexed:

```python
import numpy as np
import torch
from pathlib import Path
from ultralytics.models.yolo.detect.val import DetectionValidator

val = DetectionValidator(args=dict(plots=False))
val.metrics.names = {0: 'class0', 1: 'class1'}  # optional

# Replace this loop with your IO loading logic
for gt, pred in iterable_of_your_samples:
    # gt: dict with 'cls': [M], 'bboxes': [[x1,y1,x2,y2], ...]
    # pred: dict with 'cls': [N], 'conf': [N], 'bboxes': [[x1,y1,x2,y2], ...]
    pbatch = {
        'cls': torch.tensor(gt['cls'], dtype=torch.int64),
        'bboxes': torch.tensor(gt['bboxes'], dtype=torch.float32)
    }
    predn = {
        'cls': torch.tensor(pred['cls'], dtype=torch.int64),
        'conf': torch.tensor(pred['conf'], dtype=torch.float32),
        'bboxes': torch.tensor(pred['bboxes'], dtype=torch.float32)
    }
    no_pred = predn['cls'].numel() == 0
    stat = {
        **val._process_batch(predn, pbatch),  # returns 'tp' (and others for seg/pose)
        'target_cls': pbatch['cls'].cpu().numpy(),
        'target_img': np.unique(pbatch['cls'].cpu().numpy()),
        'conf': np.zeros(0) if no_pred else predn['conf'].cpu().numpy(),
        'pred_cls': np.zeros(0) if no_pred else predn['cls'].cpu().numpy(),
    }
    val.metrics.update_stats(stat)

# Aggregate over the entire dataset and compute AP
val.metrics.process(save_dir=Path('.'), plot=False)
print(val.metrics.results_dict)           # dict with mAP50, mAP50-95, etc.
print(val.metrics.mean_results())         # [mp, mr, mAP50, mAP50-95]
```

If anything in your saved format differs (e.g., normalized coords), make them consistent before calling `_process_batch()`.

**Comment by cavalier501** [comment]
> Glad it helped—confirming: AP is computed once at the end when you call `DetMetrics.process()`, which aggregates all per-image stats; you can then read `metrics.results_dict` or `metrics.mean_results()` for mp, mr, mAP50, mAP50-95. See the method details in the [DetMetrics reference](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.DetMetrics.process) and how stats are populated in [DetectionValidator.update_metrics](https://docs.ultralytics.com/reference/models/yolo/detect/val/#ultralytics.models.yolo.detect.val.DetectionValidator.update_metrics).
> 
> Here’s a minimal offline example for 8.3.174 using your saved preds/labels; ensure GT and prediction boxes are in the same coordinate frame (XYXY), and classes are zero-indexed:
> 
> import numpy as np
> import torch
> from pathlib import Path
> from ultralytics.models.yolo.detect.val import DetectionValidator
> 
> val = DetectionValidator(args=dict(plots=False))
> val.metrics.names = {0: 'class0', 1: 'class1'}  # optional
> 
> # Replace this loop with your IO loading logic
> for gt, pred in iterable_of_your_samples:
>     # gt: dict with 'cls': [M], 'bboxes': [[x1,y1,x2,y2], ...]
>     # pred: dict with 'cls': [N], 'conf': [N], 'bboxes': [[x1,y1,x2,y2], ...]
>     pbatch = {
>         'cls': torch.tensor(gt['cls'], dtype=torch.int64),
>         'bboxes': torch.tensor(gt['bboxes'], dtype=torch.float32)
>     }
>     predn = {
>         'cls': torch.tensor(pred['cls'], dtype=torch.int64),
>         'conf': torch.tensor(pred['conf'], dtype=torch.float32),
>         'bboxes': torch.tensor(pred['bboxes'], dtype=torch.float32)
>     }
>     no_pred = predn['cls'].numel() == 0
>     stat = {
>         **val._process_batch(predn, pbatch),  # returns 'tp' (and others for seg/pose)
>         'target_cls': pbatch['cls'].cpu().numpy(),
>         'target_img': np.unique(pbatch['cls'].cpu().numpy()),
>         'conf': np.zeros(0) if no_pred else predn['conf'].cpu().numpy(),
>         'pred_cls': np.zeros(0) if no_pred else predn['cls'].cpu().numpy(),
>     }
>     val.metrics.update_stats(stat)
> 
> # Aggregate over the entire dataset and compute AP
> val.metrics.process(save_dir=Path('.'), plot=False)
> print(val.metrics.results_dict)           # dict with mAP50, mAP50-95, etc.
> print(val.metrics.mean_results())         # [mp, mr, mAP50, mAP50-95]
> If anything in your saved format differs (e.g., normalized coords), make them consistent before calling `_process_batch()`.

Got it. Thanks for your reply much.

**Comment by glenn-jocher** [comment]
Glad it’s resolved—if you also need per-class results offline, call `metrics.summary(decimals=4)` after `metrics.process()`, which returns one row per class with P/R/F1, mAP50 and mAP50‑95; see [DetMetrics.summary](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.DetMetrics.summary).

**Comment by chinnl** [comment]
> Glad it helped—confirming: AP is computed once at the end when you call `DetMetrics.process()`, which aggregates all per-image stats; you can then read `metrics.results_dict` or `metrics.mean_results()` for mp, mr, mAP50, mAP50-95. See the method details in the [DetMetrics reference](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.DetMetrics.process) and how stats are populated in [DetectionValidator.update_metrics](https://docs.ultralytics.com/reference/models/yolo/detect/val/#ultralytics.models.yolo.detect.val.DetectionValidator.update_metrics).
> 
> Here’s a minimal offline example for 8.3.174 using your saved preds/labels; ensure GT and prediction boxes are in the same coordinate frame (XYXY), and classes are zero-indexed:
> 
> import numpy as np
> import torch
> from pathlib import Path
> from ultralytics.models.yolo.detect.val import DetectionValidator
> 
> val = DetectionValidator(args=dict(plots=False))
> val.metrics.names = {0: 'class0', 1: 'class1'}  # optional
> 
> # Replace this loop with your IO loading logic
> for gt, pred in iterable_of_your_samples:
>     # gt: dict with 'cls': [M], 'bboxes': [[x1,y1,x2,y2], ...]
>     # pred: dict with 'cls': [N], 'conf': [N], 'bboxes': [[x1,y1,x2,y2], ...]
>     pbatch = {
>         'cls': torch.tensor(gt['cls'], dtype=torch.int64),
>         'bboxes': torch.tensor(gt['bboxes'], dtype=torch.float32)
>     }
>     predn = {
>         'cls': torch.tensor(pred['cls'], dtype=torch.int64),
>         'conf': torch.tensor(pred['conf'], dtype=torch.float32),
>         'bboxes': torch.tensor(pred['bboxes'], dtype=torch.float32)
>     }
>     no_pred = predn['cls'].numel() == 0
>     stat = {
>         **val._process_batch(predn, pbatch),  # returns 'tp' (and others for seg/pose)
>         'target_cls': pbatch['cls'].cpu().numpy(),
>         'target_img': np.unique(pbatch['cls'].cpu().numpy()),
>         'conf': np.zeros(0) if no_pred else predn['conf'].cpu().numpy(),
>         'pred_cls': np.zeros(0) if no_pred else predn['cls'].cpu().numpy(),
>     }
>     val.metrics.update_stats(stat)
> 
> # Aggregate over the entire dataset and compute AP
> val.metrics.process(save_dir=Path('.'), plot=False)
> print(val.metrics.results_dict)           # dict with mAP50, mAP50-95, etc.
> print(val.metrics.mean_results())         # [mp, mr, mAP50, mAP50-95]
> If anything in your saved format differs (e.g., normalized coords), make them consistent before calling `_process_batch()`.

Hi, I'm using Ultralytics 8.2.76 and there is a file named 'eval.py'. But it doesnt support OBB validation. I uncommented the lines for loading OBB labels (line 79-80), then evaluate the predictions saved from calling 'yolo obb val ...' . But the results from the file is not the same as running the command. 
Can you explain why the results differ, and why doesnt it support OBB validation?

**Comment by glenn-jocher** [comment]
That difference is expected: OBB validation does not follow the older `eval.py` detect path, it goes through [`OBBValidator`](https://docs.ultralytics.com/reference/models/yolo/obb/val/#ultralytics.models.yolo.obb.val.OBBValidator), which switches the task to `obb` in [`init_metrics()`](https://docs.ultralytics.com/reference/models/yolo/obb/val/#ultralytics.models.yolo.obb.val.OBBValidator.init_metrics) and matches rotated boxes with `batch_probiou()` in [`_process_batch()`](https://docs.ultralytics.com/reference/models/yolo/obb/val/#ultralytics.models.yolo.obb.val.OBBValidator._process_batch), so simply uncommenting OBB label loading in `eval.py` will not reproduce `yolo obb val`; since this is a separate OBB question on a closed detect issue, please open a new issue with a minimal repro, your `eval.py` changes, full text output, and `yolo checks` so we can look at the 8.2.76 path properly.
